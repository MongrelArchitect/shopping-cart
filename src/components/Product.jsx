import React from 'react';

export default function Product({ product }) {
  return (
    <div className="product">
      <span>{product.title}</span>
      <img src={product.image} alt="" />
      <span>{product.description}</span>
      <span>
        $
        {product.price}
      </span>
    </div>
  );
}
