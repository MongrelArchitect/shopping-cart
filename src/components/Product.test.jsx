import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Product from './Product';

describe('Product component', () => {
  const mockProduct = {
    title: 'Some Fake Product',
    image: '#',
    description: 'A good ol fashioned phony placeholder product',
    id: 22,
    price: 150,
  };

  it('renders correctly', () => {
    const { container } = render(
      <Product product={mockProduct} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('calls addToCart correct number of times', () => {
    const mockAddToCart = jest.fn();
    render(<Product product={mockProduct} addToCart={mockAddToCart} />);
    const button = screen.getByRole('button');
    userEvent.click(button);
    expect(mockAddToCart).toHaveBeenCalledTimes(1);
    userEvent.click(button);
    userEvent.click(button);
    expect(mockAddToCart).toHaveBeenCalledTimes(3);
  });

  it('calls addToCart with correct arguments', () => {
    const mockAddToCart = jest.fn();
    render(<Product product={mockProduct} addToCart={mockAddToCart} />);
    const button = screen.getByRole('button');
    userEvent.click(button);
    expect(mockAddToCart).toHaveBeenCalledWith(22, 1);
  });

  it('calls addToCart with correct arguments after modifying quantity', () => {
    const mockAddToCart = jest.fn();
    render(<Product product={mockProduct} addToCart={mockAddToCart} />);
    const input = screen.getByRole('spinbutton');
    act(() => {
      userEvent.type(input, '4');
    });
    const button = screen.getByRole('button');
    userEvent.click(button);
    expect(mockAddToCart).toHaveBeenCalledWith(22, 4);
  });

  it('does not call addToCart with invalid quantity', () => {
    const mockAddToCart = jest.fn();
    render(<Product product={mockProduct} addToCart={mockAddToCart} />);
    const input = screen.getByRole('spinbutton');
    act(() => {
      userEvent.type(input, '-4');
    });
    const button = screen.getByRole('button');
    userEvent.click(button);
    expect(mockAddToCart).not.toHaveBeenCalled();
  });
});
