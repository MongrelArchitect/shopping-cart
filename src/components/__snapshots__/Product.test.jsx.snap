// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component renders correctly 1`] = `
<div>
  <div
    class="product"
  >
    <span
      class="product-title"
    >
      Some Fake Product
    </span>
    <img
      alt=""
      src="#"
    />
    <span>
      A good ol fashioned phony placeholder product
    </span>
    <span
      class="product-price"
    >
      $
      150
    </span>
    <label
      for="quant22"
    >
      Quantity:
      <input
        id="quant22"
        min="1"
        placeholder="1"
        type="number"
      />
    </label>
    <button
      type="button"
    >
      ADD TO CART
    </button>
    <span
      class="quantity-error"
      style="visibility: hidden;"
    >
      Invalid: No decimals or negatives
    </span>
  </div>
</div>
`;
